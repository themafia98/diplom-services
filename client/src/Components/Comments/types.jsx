import PropTypes from 'prop-types';
import { udataType } from '../../types';
const { bool, object, func, string } = PropTypes;

export const commentType = {
  rules: bool.isRequired,
  it: object.isRequired,
  onDelete: func.isRequired,
  userId: string.isRequired,
  uId: string.isRequired,
};

export const commentsContainerType = {
  rules: bool.isRequired,
  onUpdate: func.isRequired,
  data: object.isRequired,
  udata: udataType.isRequired,
};
