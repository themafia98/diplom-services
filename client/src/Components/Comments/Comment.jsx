import React, { useState } from "react";
import PropTypes from "prop-types";
import moment from "moment";

const Comment = ({ onDelete, rules, it }) => {
    const [key] = useState(it.id ? it.id : Math.random());

    const onDeleteEvent = event => {
        onDelete(event, key);
    };

    return (
        <p className="block-comment">
            {rules ? (
                <span className="commentControllers">
                    <span className="editComment icon-edit"></span>
                    <span onClick={onDeleteEvent} className="deleteComment icon-trash-empty"></span>
                </span>
            ) : null}
            <span className="aboutCommentSender">
                <span className="timeComment">&nbsp;{moment(it.time).format("DD.MM.YYYY HH:mm")}.</span>
                &nbsp;<span className="sender_name">{`${it.username}`}</span> написал:
            </span>
            <span className="commentContet">{it.message}</span>
        </p>
    );
};

Comment.propTypes = {
    rules: PropTypes.bool.isRequired,
    it: PropTypes.object.isRequired,
    onDelete: PropTypes.func.isRequired,
};

export default Comment;
