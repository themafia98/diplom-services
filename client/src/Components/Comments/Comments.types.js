import PropTypes from 'prop-types';
import { udataType } from 'App.types';
const { bool, object, func, string, oneOfType, number, shape, array } = PropTypes;

export const commentType = {
  rules: bool.isRequired,
  it: object.isRequired,
  userId: string.isRequired,
  router: object.isRequired,
  uId: oneOfType([number, string]).isRequired,
  onDelete: func.isRequired,
  onEdit: func,
  removeTab: func,
  onOpenPageWithData: func,
  setCurrentTab: func,
};

export const commentsContainerType = {
  rules: bool,
  onUpdate: func,
  data: shape({
    comments: array,
    _id: string,
    key: string,
  }).isRequired,
};
