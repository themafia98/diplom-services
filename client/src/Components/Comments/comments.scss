.comments {
    display: flex;
    align-items: center;
    flex-direction: column;
    justify-content: center;
    .comments__controllers {
        width: 75%;
        .sendCommentsButton {
            margin-left: 5px;
            margin-top: 3px;
        }
    }
    .commnetsListBox {
        width: 90%;
        min-width: 85%;
        min-height: 200px;
        border: 1px solid $blueDark;
        padding: 5px;
        margin-bottom: 10px;
        height: 300px;
        overflow: hidden;
        .aboutCommentSender {
            font-weight: bold;
            .sender_name {
                cursor: pointer;
                color: $blueDark;
            }
        }
        .commentContet {
            display: flex;
            width: 98%;
            border-bottom: 1px solid black;
        }
    }
}
