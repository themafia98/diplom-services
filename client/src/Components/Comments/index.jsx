import React from "react";
import moment from "moment";
import { Input, Button } from "antd";

const { TextArea } = Input;

class Comments extends React.Component {
    render() {
        return (
            <div className="comments">
                <div className="commnetsListBox">
                    <p>
                        <span className="aboutCommentSender">
                            <span className="timeComment">&nbsp;{moment().format("L")}.</span>
                            &nbsp;<span className="sender_name">Павел Петрович</span> написал:
                        </span>
                        <span className="commentContet">
                            Super Task Super Task Super Task Super Task Super Task Super Task Super Task Super Task
                            Super Task Super Task
                        </span>
                    </p>
                </div>
                <div className="comments__controllers">
                    <TextArea rows={4} />
                    <Button className="sendCommentsButton" type="primary">
                        Опубликовать
                    </Button>
                </div>
            </div>
        );
    }
}
export default Comments;
