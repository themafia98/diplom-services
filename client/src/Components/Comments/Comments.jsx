import React, { memo, useMemo, useState } from 'react';
import { commentsContainerType } from './Comments.types';
import Scrollbars from 'react-custom-scrollbars';
import _ from 'lodash';
import { v4 as uuid } from 'uuid';
import moment from 'moment';
import { Button, Empty, message, notification } from 'antd';

import Textarea from 'Components/Textarea';
import Comment from './Comment';
import actionsTypes from 'actions.types';
import { routeParser } from 'Utils';
import { withClientDb } from 'Models/ClientSideDatabase';

const Comments = memo(({ data, onUpdate, path, udata, clientDB, rules, commentProps }) => {
  const [updateDisabled, setDisable] = useState(false);
  const [msg, setValue] = useState(null);

  const { comments, key, _id: id } = data;
  const { displayName = '', _id: uId = '' } = udata;

  const addCommentsDelay = useMemo(
    () =>
      _.debounce(async () => {
        try {
          if (!msg) return message.error('Вы ничего не ввели.');
          else if (!(key && Array.isArray(comments) && !_.isEmpty(data))) return;

          const time = moment().format('DD.MM.YYYY HH:mm');

          if (!uId || !time || !displayName) {
            throw new Error('Invalid data');
          }

          const comment = {
            id: uuid(),
            uId,
            time,
            username: displayName,
            message: msg,
          };

          setDisable(true);
          setValue('');

          const parsedRoutePath = routeParser({
            pageType: 'moduleItem',
            path,
          });

          await onUpdate({
            actionType: actionsTypes.$UPDATE_SINGLE,
            parsedRoutePath,
            key,
            id,
            updateBy: '_id',
            updateItem: [...comments, comment],
            updateField: 'comments',
            store: 'tasks',
            clientDB,
            systemMessage: { done: 'Коментарий добавлен.', error: 'Комментарий добавить не удалось' },
          });

          setDisable(false);
          return;
        } catch (error) {
          console.error(error);
          notification.error({ message: 'Ошибка', description: 'Некоректные данные.' });
          setDisable(false);
        }
      }, 500),
    [clientDB, comments, data, id, key, msg, onUpdate, path, displayName, uId],
  );

  const addComments = (event) => {
    const { which, keyCode } = event;
    const isDefiend = which && keyCode;

    if (isDefiend && (which || keyCode) !== 13) return;

    if (isDefiend && (which || keyCode) === 13) {
      event.preventDefault();
    }

    addCommentsDelay(event);
  };

  const onDelete = async (event, idComment) => {
    try {
      const filterComments = comments.filter(({ id = '' }) => id !== idComment);

      const parsedRoutePath = routeParser({
        pageType: 'moduleItem',
        path,
      });

      await onUpdate({
        actionType: actionsTypes.$UPDATE_SINGLE,
        parsedRoutePath,
        id,
        key,
        item: data,
        store: 'tasks',
        updateBy: '_id',
        updateItem: filterComments,
        updateField: 'comments',
        clientDB,
        systemMessage: { done: 'Коментарий успешно удален.', error: 'Комментарий удалить не удалось' },
      });
    } catch (error) {
      console.error(error);
      notification.error({ message: 'Ошибка', description: 'Некорректная операция.' });
    }
  };

  const onEdit = async (idComment, msg = null, callback = null) => {
    if (typeof msg !== 'string') {
      message.error('Сообщение не валидно.');
      if (callback) callback();
      return;
    }

    const commentIndex = comments.findIndex((item) => item?.id === idComment);
    if (commentIndex === -1) {
      message.error('Коментарий не существует.');
      if (callback) callback();
      return;
    }
    const newCommentsArray = [...comments];
    newCommentsArray[commentIndex] = { ...comments[commentIndex], message: msg };

    try {
      const parsedRoutePath = routeParser({
        pageType: 'moduleItem',
        path,
      });

      await onUpdate({
        actionType: actionsTypes.$UPDATE_SINGLE,
        parsedRoutePath,
        id,
        key,
        item: data,
        store: 'tasks',
        updateBy: '_id',
        updateItem: newCommentsArray,
        updateField: 'comments',
        clientDB,
        systemMessage: { done: 'Коментарий успешно обновлен.', error: 'Комментарий обновить не удалось' },
      });

      if (callback) callback();
    } catch (error) {
      if (callback) callback();
      console.error(error);
      notification.error({ message: 'Ошибка', description: 'Некорректная операция.' });
    }
  };

  const onChange = ({ target }) => {
    const { value } = target;
    setValue(value);
  };

  const renderComments = (commentsArray) => {
    return commentsArray?.length ? (
      commentsArray.map((it) => (
        <Comment
          key={it.id}
          rules={rules}
          it={it}
          userId={uId}
          onDelete={onDelete}
          onEdit={onEdit}
          {...commentProps}
        />
      ))
    ) : (
      <Empty description={<span>Данных нету</span>} />
    );
  };

  return (
    <div className="comments">
      <div className="commnetsListBox">
        <Scrollbars autoHide hideTracksWhenNotNeeded>
          {renderComments(comments)}
        </Scrollbars>
      </div>
      <div className="comments__controllers">
        <Textarea
          className="comments_textarea_fild"
          onKeyDown={addComments}
          key="comments_textarea_fild"
          value={msg}
          onChange={onChange}
          rows={4}
        />
        <Button
          onClick={addComments}
          disabled={updateDisabled}
          loading={updateDisabled}
          className="sendCommentsButton"
          type="primary"
        >
          Опубликовать
        </Button>
      </div>
    </div>
  );
});

Comments.propTypes = commentsContainerType;

Comments.defaultProps = {
  data: null,
  onUpdate: null,
  rules: false,
};

export default withClientDb(Comments);
