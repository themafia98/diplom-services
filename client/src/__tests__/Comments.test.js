import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Comments from "../Components/Comments/index";

it("Comments test", () => {
    const props = {
        rules: true,
        data: {
            key: "sad",
            comments: [
                {
                    id: "213123",
                    time: "DD.MM.YYYY HH:mm",
                    username: "Павел Петрович",
                    message: "sadasdasdas",
                },
            ],
        },
        onUpdate: () => {},
    };
    const CommentsWrapper = shallow(<Comments {...props} />);
    expect(toJson(CommentsWrapper)).toMatchSnapshot();

    CommentsWrapper.setState({ onUpdateDisabled: true });
    expect(CommentsWrapper.find(".sendCommentsButton").prop("disabled")).toEqual(true);
    expect(CommentsWrapper).toMatchSnapshot();

    CommentsWrapper.setState({ onUpdateDisabled: false });
    expect(CommentsWrapper.find(".sendCommentsButton").prop("disabled")).toEqual(false);
    expect(CommentsWrapper).toMatchSnapshot();

    expect(CommentsWrapper.find(".sendCommentsButton").prop("disabled")).toEqual(false);

    expect(CommentsWrapper.find(".sendCommentsButton").simulate("click", []));
    expect(CommentsWrapper).toMatchSnapshot();

    expect(CommentsWrapper.find("Comment")).toHaveLength(1);
});
