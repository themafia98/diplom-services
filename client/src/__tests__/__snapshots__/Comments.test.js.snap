// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentView /> Should render component 1`] = `
<div
  className="comments"
>
  <div
    className="commnetsListBox"
  >
    <Scrollbars
      autoHeight={false}
      autoHeightMax={200}
      autoHeightMin={0}
      autoHide={false}
      autoHideDuration={200}
      autoHideTimeout={1000}
      hideTracksWhenNotNeeded={false}
      renderThumbHorizontal={[Function]}
      renderThumbVertical={[Function]}
      renderTrackHorizontal={[Function]}
      renderTrackVertical={[Function]}
      renderView={[Function]}
      tagName="div"
      thumbMinSize={30}
      universal={false}
    >
      <Comment
        it={
          Object {
            "id": "213123",
            "message": "sadasdasdas",
            "time": "DD.MM.YYYY HH:mm",
            "username": "Павел Петрович",
          }
        }
        key="213123"
        onDelete={[Function]}
        rules={true}
      />
    </Scrollbars>
  </div>
  <div
    className="comments__controllers"
  >
    <Textarea
      className="comments_textarea_fild"
      key="comments_textarea_fild"
      onChange={[Function]}
      onKeyDown={[Function]}
      rows={4}
      value={null}
    />
    <Button
      block={false}
      className="sendCommentsButton"
      disabled={false}
      ghost={false}
      htmlType="button"
      loading={false}
      onClick={[Function]}
      type="primary"
    >
      Опубликовать
    </Button>
  </div>
</div>
`;

exports[`<ContentView /> Should render component 2`] = `
<div
  className="comments"
>
  <div
    className="commnetsListBox"
  >
    <Scrollbars
      autoHeight={false}
      autoHeightMax={200}
      autoHeightMin={0}
      autoHide={false}
      autoHideDuration={200}
      autoHideTimeout={1000}
      hideTracksWhenNotNeeded={false}
      renderThumbHorizontal={[Function]}
      renderThumbVertical={[Function]}
      renderTrackHorizontal={[Function]}
      renderTrackVertical={[Function]}
      renderView={[Function]}
      tagName="div"
      thumbMinSize={30}
      universal={false}
    >
      <Comment
        it={
          Object {
            "id": "213123",
            "message": "sadasdasdas",
            "time": "DD.MM.YYYY HH:mm",
            "username": "Павел Петрович",
          }
        }
        key="213123"
        onDelete={[Function]}
        rules={true}
      />
    </Scrollbars>
  </div>
  <div
    className="comments__controllers"
  >
    <Textarea
      className="comments_textarea_fild"
      key="comments_textarea_fild"
      onChange={[Function]}
      onKeyDown={[Function]}
      rows={4}
      value={null}
    />
    <Button
      block={false}
      className="sendCommentsButton"
      disabled={true}
      ghost={false}
      htmlType="button"
      loading={true}
      onClick={[Function]}
      type="primary"
    >
      Опубликовать
    </Button>
  </div>
</div>
`;

exports[`<ContentView /> Should render component 3`] = `
<div
  className="comments"
>
  <div
    className="commnetsListBox"
  >
    <Scrollbars
      autoHeight={false}
      autoHeightMax={200}
      autoHeightMin={0}
      autoHide={false}
      autoHideDuration={200}
      autoHideTimeout={1000}
      hideTracksWhenNotNeeded={false}
      renderThumbHorizontal={[Function]}
      renderThumbVertical={[Function]}
      renderTrackHorizontal={[Function]}
      renderTrackVertical={[Function]}
      renderView={[Function]}
      tagName="div"
      thumbMinSize={30}
      universal={false}
    >
      <Comment
        it={
          Object {
            "id": "213123",
            "message": "sadasdasdas",
            "time": "DD.MM.YYYY HH:mm",
            "username": "Павел Петрович",
          }
        }
        key="213123"
        onDelete={[Function]}
        rules={true}
      />
    </Scrollbars>
  </div>
  <div
    className="comments__controllers"
  >
    <Textarea
      className="comments_textarea_fild"
      key="comments_textarea_fild"
      onChange={[Function]}
      onKeyDown={[Function]}
      rows={4}
      value={null}
    />
    <Button
      block={false}
      className="sendCommentsButton"
      disabled={false}
      ghost={false}
      htmlType="button"
      loading={false}
      onClick={[Function]}
      type="primary"
    >
      Опубликовать
    </Button>
  </div>
</div>
`;

exports[`<ContentView /> Should render component 4`] = `
<div
  className="comments"
>
  <div
    className="commnetsListBox"
  >
    <Scrollbars
      autoHeight={false}
      autoHeightMax={200}
      autoHeightMin={0}
      autoHide={false}
      autoHideDuration={200}
      autoHideTimeout={1000}
      hideTracksWhenNotNeeded={false}
      renderThumbHorizontal={[Function]}
      renderThumbVertical={[Function]}
      renderTrackHorizontal={[Function]}
      renderTrackVertical={[Function]}
      renderView={[Function]}
      tagName="div"
      thumbMinSize={30}
      universal={false}
    >
      <Comment
        it={
          Object {
            "id": "213123",
            "message": "sadasdasdas",
            "time": "DD.MM.YYYY HH:mm",
            "username": "Павел Петрович",
          }
        }
        key="213123"
        onDelete={[Function]}
        rules={true}
      />
    </Scrollbars>
  </div>
  <div
    className="comments__controllers"
  >
    <Textarea
      className="comments_textarea_fild"
      key="comments_textarea_fild"
      onChange={[Function]}
      onKeyDown={[Function]}
      rows={4}
      value={null}
    />
    <Button
      block={false}
      className="sendCommentsButton"
      disabled={false}
      ghost={false}
      htmlType="button"
      loading={false}
      onClick={[Function]}
      type="primary"
    >
      Опубликовать
    </Button>
  </div>
</div>
`;
